### DISCUSSION TITLE
Tauri vs Electron for use with and MonacoLanguageClient

### DISCUSSION THREAD

**Discussion by Sahasrara** [TypeFox/monaco-languageclient#570] [Q&A] [answered]
I'm going to start writing a desktop app (non-linear dialogue editor for games) using either Tauri or Electron.

I'd like to make use of Monaco Editor and I need language support for C++/C#/GDScript/etc. These languages already have LSP implementations that I can launch as child processes from either Tauri or Electron.

Either framework can spawn child processes and communicate with them view stdout/stdin/stderr.

My questions are:
1. In order to use MonacoLanguageClient, would I need to implement AbstractMessageReader and AbstractMessageWriter as StdinMessageWriter/StdoutMessageReader and then initialize MonacoLanguageClient with those inside a MessageTransports container object?
2. Is there a better way to do this? For instance, is there some way to run existing VSCode extensions on Electron?
3. Is there any reason you would recommend Tauri or Electron over the other?

**Comment by kaisalmen** [comment, marked as answer]
Hi @Sahasrara I saw this, did not answer immediately and forgot. But here comes the answer.

> 1. In order to use MonacoLanguageClient, would I need to implement AbstractMessageReader and AbstractMessageWriter as StdinMessageWriter/StdoutMessageReader and then initialize MonacoLanguageClient with those inside a MessageTransports container object?

`monaco-languageclient` is usually needed where you cannot write to std-in/std-out. In a browser context this is not possible. That's why in our examples with a non web-worker language server (pyhton, json) a node app serves as translator like this:
**browser** <-websocket-> **node-app** <-std-in/std-out-> **language-server**

If you have access from your client/front-end directly to the file-system (like in vscode desktop version) this library is not needed as both can use this interaction mechanism. You can then directly use the underlying `vscode-languageclient`

> 2. Is there a better way to do this? For instance, is there some way to run existing VSCode extensions on Electron?

Electron apps are able to access the file system, so it could/should be possible to directly run the client extension with `monaco-vscode-api`, but we don't have any examples for Electron not in this repo nor [there](https://github.com/CodinGame/monaco-vscode-api)

> 3. Is there any reason you would recommend Tauri or Electron over the other?

I can't give you a qualified recommendation due to lack of experience with those technologies from my side.

**Reply by CGNonofr** [reply]
Btw, monaco-vscode-api uses web services of VSCode, so running it in a node environment will not provide the same experience as in VSCode desktop. I'm currently thinking on how to integrate desktop services as an alternative
